clio-client/src/main/scala/org/broadinstitute/clio/client/ClioClient.scala

```diff
@@ -1,61 +1,259 @@
-import caseapp._
+import caseapp.core.{Messages, WithHelp}
 import com.typesafe.scalalogging.LazyLogging
-import org.broadinstitute.clio.client.commands.Commands._
-import org.broadinstitute.clio.client.commands.{CommandDispatch, CommandType}
-import org.broadinstitute.clio.client.util.IoUtil
-import org.broadinstitute.clio.client.webclient.ClioWebClient
 
+import scala.concurrent.{ExecutionContext, Future}
 import scala.util.{Failure, Success}
 
-object ClioClient
-    extends CommandAppWithPreCommand[CommonOptions, CommandType]
-    with LazyLogging {
-  override val progName = "clio-client"
-  override val appName = "Clio Client"
+/**
+  * The command-line entry point for the clio-client.
+  *
+  * We want most actual logic to happen inside this
+  * object's companion class for maximum testability,
+  * so this object should only handle:
+  *
+  *   1. Setting up and tearing down the actor system
+  *      infrastructure needed by the client.
+  *
+  *   2. Exiting the program with appropriate return
+  *      codes.
+  */
+object ClioClient {
+
+  /**
+    * Model representing termination of the client before
+    * it even gets to running a sub-command.
+    *
+    * This could be because of an input error, or because
+    * the user asked for help / usage.
+    *
+    * We model the two cases differently so we can exit with
+    * different return codes accordingly.
+    */
+  sealed trait EarlyReturn
+  final case class ParsingError(error: String) extends EarlyReturn
+  final case class UsageOrHelpAsked(message: String) extends EarlyReturn
 
-  override val appVersion: String = ClioClientConfig.Version.value
+  val progName = "clio-client"
 
-  //not sure this is the best way to pass common opts but case-app docs are spotty at best
-  var commonOptions: CommonOptions = _
+  def main(args: Array[String]): Unit = {
+    implicit val system: ActorSystem = ActorSystem(progName)
+    import system.dispatcher
+    sys.addShutdownHook({ val _ = system.terminate() })
 
-  override def beforeCommand(options: CommonOptions,
-                             remainingArgs: Seq[String]): Unit = {
-    commonOptions = options
-    checkRemainingArgs(remainingArgs)
+    val webClient: ClioWebClient = new ClioWebClient(
+      ClioClientConfig.ClioServer.clioServerHostName,
+      ClioClientConfig.ClioServer.clioServerPort,
+      ClioClientConfig.ClioServer.clioServerUseHttps
+    )
+    val client: ClioClient = new ClioClient(webClient, IoUtil)
+
+    client
+      .instanceMain(args)
+      .fold(
+        {
+          case UsageOrHelpAsked(message) => {
+            println(message)
+            Future.successful(())
+          }
+          case ParsingError(error) => {
+            System.err.println(error)
+            Future.failed(new RuntimeException)
+          }
+        },
+        identity
+      )
+      .onComplete {
+        case Success(_) => sys.exit(0)
+        case Failure(_) => sys.exit(1)
```

Comment:
We should be logging errors on failure.

Reply:
Yes, and in general we should be logging errors at the top level, not logging at the level where the error is first generated.